src/binarymanager.ts

```diff
@@ -0,0 +1,124 @@
+import axios, { AxiosResponse } from 'axios';
+import which from 'which';
+import { globalContext } from './extension';
+import { MirrordAPI } from './api';
+import { Utils } from 'vscode-uri';
+import * as fs from 'node:fs';
+import { Uri, workspace, window, ProgressLocation } from 'vscode';
+
+const mirrordBinaryEndpoint = 'https://version.mirrord.dev/v1/version';
+const binaryCheckInterval = 1000 * 60 * 3;
+
+
+function getExtensionMirrordPath(): Uri {
+    return Utils.joinPath(globalContext.globalStorageUri, 'mirrord');
+}
+
+
+/**
+ * Returns path to mirrord binary
+ */
+export async function getMirrordBinaryPath(): Promise<string> {
+    const extensionMirrordPath = getExtensionMirrordPath();
+
+    const latestVersion = await getLatestSupportedVersion();
+
+    // See if maybe we have it installed already, in correct version.
+    try {
+        const mirrordPath = await which("mirrord");
+        const api = new MirrordAPI(mirrordPath);
+        const installedVersion = await api.getBinaryVersion();
+        if (installedVersion === latestVersion) {
+            return mirrordPath;
+        }
+    } catch (e) {
+        // don't care
+    }
+
+    // Check if we previously installed latest version.
+    let binaryExists = false;
+    try {
+        await workspace.fs.stat(extensionMirrordPath);
+        binaryExists = true;
+    } catch (e) {
+        // that's okay
+    }
+
+    if (binaryExists) {
+        const api = new MirrordAPI(extensionMirrordPath.fsPath);
+        const installedVersion = await api.getBinaryVersion();
+        if (installedVersion === latestVersion) {
+            return extensionMirrordPath.fsPath;
+        }
+    }
+
+    await downloadMirrordBinary(extensionMirrordPath, latestVersion);
+
+    return extensionMirrordPath.fsPath;
+}
+
+/**
+ * 
+ * @returns The latest supported version of mirrord for current extension version
+ */
+async function getLatestSupportedVersion(): Promise<string> {
+    let lastChecked = globalContext.globalState.get('binaryLastChecked', 0);
+    let lastBinaryVersion = globalContext.globalState.get('lastBinaryVersion', '');
+
+    if (lastBinaryVersion && lastChecked > Date.now() - binaryCheckInterval) {
```

Comment:
IMO lets remove the check interval - have it only apply for how often we display the message